### DISCUSSION TITLE
How to handle chain-aware Alchemy API keys ?

### DISCUSSION THREAD

**Discussion by AtlasRW** [wevm/wagmi#974] [Q&A]
Hello everyone,

I'm having trouble while configuring my Alchemy Provider because Alchemy now uses specific API keys per chain, so providing an API key created for Polygon Mumbai should not work when on Polygon Mainnet, I suppose.

**Comment by tmm** [comment]
Unless something changed recently, I'm fairly certain that Alchemy API keys work across chains. Do you have an example to try out?

**Reply by AtlasRW** [reply]
Thanks for your answer,

Well, when you create a new app (and its set of API keys) on Alchemy's dashboard, you have to set a specific network, so common sense led me to think it would be restricted to this network

**Reply by tmm** [reply]
Confusing UX on Alchemy's part. AFAIK they aren't restricted to a specific network.

**Comment by qqqllppp** [comment]
I had the same concern here so I decided to do a simple test.

I used an address which had some test MATIC and ETH in mumbai and goerli network respectively.

I combined Alchemy's RPC URL with different APIs generated from creating separate projects on mumbai and goerli, followed by printing the balance to compare with the balance in my wallet.

The result is that Alchemy's API keys will work cross chain (as @tmm mentioned), and what determines which chain you are on is Alchemy's RPC URL. Since wagmi's config just uses the API key, your app interaction through wagmi should be cross chain compatible.

Here is the script:

```
require("dotenv").config();
const ethers = require("ethers");

const ADDRESS = process.env.ADDRESS;

const RPC_URL_MUMBAI = "https://polygon-mumbai.g.alchemy.com/v2/";
const RPC_API_MUMBAI = process.env.RPC_API_MUMBAI; // got this from project created with "mumbai" network selected

const RPC_URL_GOERLI = "https://eth-goerli.g.alchemy.com/v2/";
const RPC_API_GOERLI = process.env.RPC_API_GOERLI; // got this from project created with "goerli" network selected

const V1 = RPC_URL_MUMBAI.concat(RPC_API_MUMBAI);
const V2 = RPC_URL_MUMBAI.concat(RPC_API_GOERLI);
const V3 = RPC_URL_GOERLI.concat(RPC_API_MUMBAI);
const V4 = RPC_URL_GOERLI.concat(RPC_API_GOERLI);

async function main() {
  const provider1 = new ethers.providers.StaticJsonRpcProvider(V1);
  const balance1 = await provider1.getBalance(ADDRESS);
  console.log(ethers.utils.formatEther(balance1));

  const provider2 = new ethers.providers.StaticJsonRpcProvider(V2);
  const balance2 = await provider2.getBalance(ADDRESS);
  console.log(ethers.utils.formatEther(balance2));

  const provider3 = new ethers.providers.StaticJsonRpcProvider(V3);
  const balance3 = await provider3.getBalance(ADDRESS);
  console.log(ethers.utils.formatEther(balance3));

  const provider4 = new ethers.providers.StaticJsonRpcProvider(V4);
  const balance4 = await provider4.getBalance(ADDRESS);
  console.log(ethers.utils.formatEther(balance4));
}

main().catch((error) => {
  console.error(error);
  process.exitCode = 1;
});
```

**Comment by basememara** [comment]
This seems like a loose assumption, plus what if someone does actually want to segment their API keys so one is assigned per chain for analytics or Alchemy app configurations. What do you think of something like this:

```typescript
const { chains, provider } = configureChains(
  [polygon, polygonMumbai],
  [
    alchemyProvider({
      apiKey: {
        matic: process.env.ALCHEMY_API_KEY_POLYGON_MAINNET,
        maticmum: process.env.ALCHEMY_API_KEY_POLYGON_MUMBAI
      }
    }),
    publicProvider()
  ]
);
```

**Reply by tmm** [reply]
Right now, you can use the [`jsonRpcProvider`](https://wagmi.sh/react/providers/jsonRpc) to achieve this behavior. We'll be addressing this more holistically in the first major version of wagmi.
